src/Resolver/Factory.php

```diff
@@ -109,12 +112,56 @@ private function decorateHostsFileExecutor(ExecutorInterface $executor)
     private function createExecutor($nameserver, LoopInterface $loop)
     {
         if ($nameserver instanceof Config) {
-            $nameserver = \reset($nameserver->nameservers);
-            if ($nameserver === false) {
+            if (!$nameserver->nameservers) {
                 throw new \UnderflowException('Empty config with no DNS servers');
             }
+
+            // Hard-coded to check up to 3 DNS servers to match default limits in place in most systems (see MAXNS config).
+            // Note to future self: Recursion isn't too hard, but how deep do we really want to go?
```

Comment:
We could respect `MAXNS` in the future, or just try them all. Just not sure if we should try them all as that is unexpected behavior when parsing resolve.conf

Reply:
Fair point, that's also why I've added this remark. I'd rather get this in as is, as it mimics common system behavior and I would rather not have to refactor this piece of logic. I've actually started with a recursive implementation, but figured the resulting code was very hard to debug (it's somewhat non-trivial in its current state already). Let's leave this up for the future? :shipit:

Reply:
Agreed, that comment wasn't meant to get it in now. But more a thought dump for the future :smile: .